### PR TITLE
KOA-4818 Add basic support for publishing compose package

### PR DESCRIPTION
**Pull request by marianeum**
This adds a very basic support for being able to publish a compose package. This is just a temporary solution until we migrate away from JS for publishing (aka terrible code ™️)

+ [x] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-android/blob/main/CONTRIBUTING.md)

Remember to include the following changes:
+ [x] `UNRELEASED.md`
+ [x] `README.md`
+ [x] Tests
+ [x] Docs (either update [backpack-docs](https://github.com/Skyscanner/backpack-docs) now, or create a follow up ticket)

_If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/main/CODE_REVIEW_GUIDELINES.md)_

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -127,7 +127,7 @@ jobs:
           api-level: ${{ env.api-level }}
           force-avd-creation: false
           emulator-options: -no-snapshot-save -no-window -gpu swiftshader_indirect -noaudio -no-boot-anim -camera-back none
-          script: ./gradlew :Backpack:connected${{ env.config }}AndroidTest
+          script: ./gradlew :Backpack:connected${{ env.config }}AndroidTest :backpack-compose:connected${{ env.config }}AndroidTest
```
Comment (marianeum) [inline .github/workflows/ci.yml:130]: make sure we also build the compose tests on CI (there are none right now, but there will be in the future)

File: backpack-common/src/main/res/values/colors.xml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+  <color name="bpkTextPrimary">#000000</color>
```
Comment (marianeum) [inline backpack-common/src/main/res/values/colors.xml:3]: this is... annoying. i found that when running the compose tests that there was a compilation error. we refer to the primary colour in the icons. in compose tinting will happen elsewhere. in the old view system this will get overridden in the other module

File: backpack-compose/build.gradle.kts [hunk 3]
```diff
@@ -20,6 +20,12 @@ plugins {
   kotlin("android")
 }
 
+ext {
```
Comment (marianeum) [inline backpack-compose/build.gradle.kts:23]: apply publishing in compose

File: backpack-compose/build.gradle.kts [hunk 4]
```diff
@@ -29,6 +35,13 @@ android {
   composeOptions {
     kotlinCompilerExtensionVersion = rootProject.ext.get("compose_version") as String
   }
+
+  packagingOptions {
```
Comment (marianeum) [inline backpack-compose/build.gradle.kts:39]: this was also needed for running the tests

File: package.json [hunk 5]
```diff
@@ -70,7 +70,7 @@
     "husky": "^7.0.0",
     "inquirer": "^8.1.2",
     "js-yaml": "^4.1.0",
-    "lint-staged": "^12.0.2",
+    "lint-staged": "^11.2.6",
```
Comment (marianeum) [inline package.json:73]: had to downgrade this as v12 is broken

File: package.json [hunk 6]
```diff
@@ -9,7 +9,7 @@
   "scripts": {
     "preinstall": "npx ensure-node-env",
     "emulator:build": "./gradlew build",
-    "emulator:instrumentation": "./gradlew Backpack:connectedDebugAndroidTest -PlocalTests=true",
+    "emulator:instrumentation": "./gradlew Backpack:connectedDebugAndroidTest backpack-compose:connectedDebugAndroidTest -PlocalTests=true",
```
Comment (marianeum) [inline package.json:12]: same as on CI - run compose tests as well. this was the easiest way to do it - we just always run the full test suite no matter what gets published

File: release.js [hunk 7]
```diff
@@ -25,39 +25,24 @@ const releaseit = require('release-it');
...
-        name: `${key} (${i[key]})`,
-        value: `${i[key]}`,
-      };
-    }),
-  },
-];
+const compose = "compose"
+const android = "android"
+
+const package_questions = [
```
Comment (marianeum) [inline release.js:35]: add a new question for what should be released

File: release.js [hunk 8]
```diff
@@ -104,17 +89,63 @@ async function amendVersionsFile(key, version) {
   }
 }
 
+async function versionPrompt(currentVersion) {
```
Comment (marianeum) [inline release.js:92]: just moved into a function

File: release.js [hunk 9]
```diff
@@ -104,17 +89,63 @@ async function amendVersionsFile(key, version) {
...
     childProcess.execSync('./gradlew :Backpack:checkMavenCredentials');
-    const { version } = await inquirer.prompt(questions);
+    const packageResult = await inquirer.prompt(package_questions);
+    const packageType = packageResult.type
+    const packageVersion = (packageType == android) ? versions.android : versions.compose
+    const { version } = await versionPrompt(packageVersion);
     const commonVersion = semver.inc(versions.common, 'minor')
 
-    await amendGradleFiles('BpkAndroidVersion', version);
+    if (packageType == android) {
```
Comment (marianeum) [inline release.js:134]: just a bunch of if checks depending on which package to publish. not ideal, but it'll do for now

File: release.js [hunk 10]
```diff
@@ -134,7 +165,8 @@ async function release() {
     await releaseit(releaseOptions);
     const publishTask = 'publishMavenPublicationToSonatypeRepository';
     const releaseTask = 'closeAndReleaseSonatypeStagingRepository';
-    childProcess.execSync(`./gradlew :backpack-common:${publishTask} :Backpack:${publishTask} ${releaseTask}`);
+    const publishModule = (packageType == android) ? 'Backpack' : 'backpack-compose'
```
Comment (marianeum) [inline release.js:168]: publish the right package depending on the type

File: release.js (hunk 9 above)
Comment (olliecurtis) [reply release.js:134]: You could always extract it to a function given it looks to be doing the same thing for each version and then you can just pass in the Name, type and version?

**Review by olliecurtis — APPROVED** [review]

File: backpack-compose/build.gradle.kts [hunk 11]
```diff
@@ -29,6 +35,13 @@ android {
...
+    resources.excludes.add("**/attach_hotspot_windows.dll")
```
Comment (zgeor) [inline backpack-compose/build.gradle.kts:40]: What is this specific DLL for?

Comment (marianeum) [reply backpack-compose/build.gradle.kts:40]: It was as part of the coroutines test dependency i think

File: package.json (hunk 5 above)
Comment (frugoman) [reply package.json:73]: I've read that from 12, we don't need to `git add` anymore on the staged task. have you tried this?

File: backpack-compose/build.gradle.kts (hunk 11 above)
Comment (zgeor) [reply backpack-compose/build.gradle.kts:40]: right I see. I went through the coroutines issue https://github.com/Kotlin/kotlinx.coroutines/issues/2023

File: package.json (hunk 5 above)
Comment (zgeor) [reply package.json:73]: This is downgrading quite a few dependencies. Likely will introduce vulnerabilities.

**Review by zgeor — APPROVED** [review]

File: package.json (hunk 5 above)
Comment (marianeum) [reply package.json:73]: This was upgraded & released very recently (few days ago), but the issue only appeared once we made changes (as lint-staged only throws errors on modified code). basically the error message is not relevant to the code, so you don't know what the lint issue is with v12 - I had a chat with @olliecurtis about this - it's to do with how we use husky to call lint-staged - in v12 it is no longer possible to do this. we'll have the same issue on all other projects
